### ISSUE TITLE
Purpose of nil bits?

### ISSUE THREAD

**Issue by coreyeng** [Origen-SDK/origen#262]
I'm back to, and almost completed, spec-ing and documenting the collector. I added some miscellaneous cases to it to make it behave closer to how a hash works, e.g., setting multiple values, setting the same value multiple times (technically this is legal, but doesn't really serve any purpose and is most likely a mistake), and empty value raise exceptions.

Currently, I know the collector accepts empty values and that empty values are used, since I printed deprecation warning at first to see if they would come in when running the full specs, and the empty case is used for registers.

Looking at the specs, its used to initialize bits. Which I can see, but is there a purpose in initializing it to nil? The bit will exist, but won't Origen treat this as a 0 when actually used and wouldn't it be clearer to initialize it to 0?

Furthermore, should the collector complain about missing values? I've only seen this case and a case in RosettaStone (internal gem) that have this issue, so it doesn't seem widely used (and, for now, it won't crash, just print deprecation warnings). However, a change to the register would require a spec change.

Since its in the spec, I have no problem keeping it, but I think an improvement would be to initialize those bits to 0, then figure out whether the collector should complain about empty values or not.

For reference, this is what I mean by empty values:

~~~ruby
Origen::Registers.default_bit_meta_data do |bit|
  bit.attr_bx
  bit.attr_by = 10
  bit.attr_bz = 20
end
~~~

Currently, I would have the collector raise an error about <code>attr_bx</code> not having a value, since as a hash <code>attr_bx: </code> would not be allowed, though <code>attr_bx nil</code> would be, as its the same as <code>attr_bx: nil</code>. That's my reasoning at least.

You can see the full scope of what I plan to PR soon here: https://github.com/coreyeng/origen/blob/master/spec/utility/collector_spec.rb

**Comment by info-rchitect** [comment]
@coreyeng did I miss the original thread on what this is used for?

**Comment by coreyeng** [comment]
Nope, there was no original thread. The collector was something @ginty put together to use mash an options hash with an options block (allows bits to be declared using either a block or has, or both).

However, I've been putting it in stuff I've been working on since you mentioned block support for Componentable back in October of last year I think. I'm spec-ing/documenting it now so its 'official' and shouldn't change on us, potentially breaking stuff that uses it but isn't itself specs (e.g., internal gems).

**Comment by coreyeng** [comment]
From the docs I'm writing:

![image](https://user-images.githubusercontent.com/13453967/39576682-9ae276de-4ea4-11e8-89cd-784700d9051e.png)

**Comment by ginty** [comment]
~~~ruby
Origen::Registers.default_bit_meta_data do |bit|
  bit.attr_bx
  bit.attr_by = 10
  bit.attr_bz = 20
end
~~~

This is not defining a bit, it is defining an attribute of that bit, and more specifically an attribute of that bit that is not its data value.
So I don't see that case for treating nil as 0 based on that example.

The collector in this case is being used to do two things, get the attribute names (the keys) and the default values of them.
It seems reasonable to be using it to define an attribute which has a default value of nil.
